dm/master/openapi.go

```diff
@@ -312,3 +377,205 @@ func modelToSourceCfg(source openapi.Source) *config.SourceConfig {
+		subTaskCfg.From = sourceCfgMap[sourceCfg.SourceName].From
+		// set target db config
+		subTaskCfg.To = *toDBCfg.Clone()
+		// TODO set meet error policy
+		// TODO ExprFilter
+		// set full unit config
+		subTaskCfg.MydumperConfig = config.DefaultMydumperConfig()
+		subTaskCfg.LoaderConfig = config.DefaultLoaderConfig()
+		if fullCfg := task.SourceConfig.FullMigrateConf; fullCfg != nil {
+			if fullCfg.ExportThreads != nil {
+				subTaskCfg.MydumperConfig.Threads = *fullCfg.ExportThreads
+			}
+			if fullCfg.ImportThreads != nil {
+				subTaskCfg.LoaderConfig.PoolSize = *fullCfg.ImportThreads
+			}
+			if fullCfg.DataDir != nil {
+				subTaskCfg.LoaderConfig.Dir = *fullCfg.DataDir
+			}
+		}
+		// set incremental config
+		subTaskCfg.SyncerConfig = config.DefaultSyncerConfig()
+		if incrCfg := task.SourceConfig.IncrMigrateConf; incrCfg != nil {
+			if incrCfg.ReplThreads != nil {
+				subTaskCfg.SyncerConfig.WorkerCount = *incrCfg.ReplThreads
+			}
+			if incrCfg.ReplBatch != nil {
+				subTaskCfg.SyncerConfig.Batch = *incrCfg.ReplBatch
+			}
+		}
+		// set route,blockAllowList,filter config
+		doDBs := []string{}
+		doTables := []*filter.Table{}
+		routeRules := []*router.TableRule{}
+		filterRules := []*bf.BinlogEventRule{}
+		for _, rule := range tableMigrateRuleMap[sourceCfg.SourceName] {
+			// route
+			routeRules = append(routeRules, &router.TableRule{
+				SchemaPattern: rule.Source.Schema, TablePattern: rule.Source.Table,
+				TargetSchema: rule.Target.Schema, TargetTable: rule.Target.Table,
+			})
+			// filter
+			if rule.EventFilterName != nil {
+				for _, name := range *rule.EventFilterName {
+					filterRule, ok := eventFilterTemplateMap[name] // NOTE: this return a copied value
+					if !ok {
+						return nil, terror.ErrOpenAPICommonError.Generatef("filter rule name=%s not found", name)
+					}
+					filterRule.SchemaPattern = rule.Source.Schema
+					filterRule.TablePattern = rule.Source.Table
+					filterRules = append(filterRules, &filterRule)
+				}
+			}
+			// BlockAllowList
+			doDBs = append(doDBs, rule.Source.Schema)
+			doTables = append(doTables, &filter.Table{Schema: rule.Source.Schema, Name: rule.Source.Table})
+		}
+		subTaskCfg.RouteRules = routeRules
+		subTaskCfg.FilterRules = filterRules
+		subTaskCfg.BAList = &filter.Rules{DoDBs: removeDuplication(doDBs), DoTables: doTables}
+		// adjust sub task config
+		if err := subTaskCfg.Adjust(true); err != nil {
+			return nil, terror.Annotatef(err, "source name=%s", sourceCfg.SourceName)
+		}
+		subTaskCfgList[i] = *subTaskCfg
+	}
+	return subTaskCfgList, nil
+}
+
+func adjustModelTask(task *openapi.Task) error {
+	defaultMetaSchema := "dm_meta"
+	if task.MetaSchema == nil {
+		task.MetaSchema = &defaultMetaSchema
+	}
+	// check some not implemented features
+	if task.OnDuplication != openapi.TaskOnDuplicationError {
+		return terror.ErrOpenAPICommonError.Generate("now on duplication is currently only implemented at the error level")
```

Comment:
```suggestion
		return terror.ErrOpenAPICommonError.Generate("`on_duplicate` only supports `error` for now.")
```

BTW the field should be `on_duplicate` not `on_duplication`

Reply:
I mean `OnDuplication` should be `OnDuplicate`

Reply:
added in https://github.com/pingcap/dm/pull/2141/commits/f965d923750be07054278f757035732c6cfc3577